R/loo_moment_matching.R

```diff
@@ -0,0 +1,568 @@
+#' update_quantities_i which updates the importance weights based on
+#' the supplied model object.
+#'
+#' @noRd
+#' @param x A fitted model object.
+#' @param upars A matrix representing a sample of vector-valued parameters in
+#' the unconstrained space
+#' @param lwi A vector representing the log-weight of each parameter
+#' @return List with the shift and scaling that were performed, and the new
+#' parameter matrix.
+#'
+#'
+shift_and_scale <- function(x, upars, lwi) {
+  # compute moments using log weights
+  S <- dim(upars)[1]
+  mean_original <- colMeans(upars)
+  mean_weighted <- colSums(exp(lwi) * upars)
+  shift <- mean_weighted - mean_original
+  mii <- exp(lwi)* upars^2
+  mii <- colSums(mii) - mean_weighted^2
+  mii <- mii*S/(S-1)
+  scaling <- sqrt(mii / matrixStats::colVars(upars))
+  # transform posterior draws
+  upars_new <- sweep(upars, 2, mean_original, "-")
+  upars_new <- sweep(upars_new, 2, scaling, "*")
+  upars_new <- sweep(upars_new, 2, mean_weighted, "+")
+
+  list(
+    upars = upars_new,
+    shift = shift,
+    scaling = scaling
+  )
+}
+
+#' Shift a matrix of parameters to their weighted mean and scale the covariance
+#' to match the weighted covariance.
+#' Also calls update_quantities_i which updates the importance weights based on
+#' the supplied model object.
+#'
+#' @noRd
+#' @param x A fitted model object.
+#' @param upars A matrix representing a sample of vector-valued parameters in
+#' the unconstrained space
+#' @param lwi A vector representing the log-weight of each parameter
+#' @return List with the shift and mapping that were performed, and the new
+#' parameter matrix.
+#'
+shift_and_cov <- function(x, upars, lwi, ...) {
+  # compute moments using log weights
+  mean_original <- colMeans(upars)
+  mean_weighted <- colSums(exp(lwi) * upars)
+  shift <- mean_weighted - mean_original
+  covv <- stats::cov(upars)
+  wcovv <- stats::cov.wt(upars, wt = exp(lwi))$cov
+  chol1 <- tryCatch(
+    {
+      chol(wcovv)
+    },
+    error = function(cond)
+    {
+      return(NULL)
+    }
+  )
+  if (is.null(chol1)) {
+    mapping <- diag(length(mean_original))
+  }
+  else {
+    chol2 <- chol(covv)
+    mapping <- t(chol1) %*% solve(t(chol2))
+  }
+  # transform posterior draws
+  upars_new <- sweep(upars, 2, mean_original, "-")
+  upars_new <- tcrossprod(upars_new, mapping)
+  upars_new <- sweep(upars_new, 2, mean_weighted, "+")
+  colnames(upars_new) <- colnames(upars)
+
+  list(
+    upars = upars_new,
+    shift = shift,
+    mapping = mapping
+  )
+}
+
+
+#' Warning message if max_iters is reached
+#' @noRd
+throw_moment_match_max_iters_warning <- function() {
+  warning(
+    "The maximum number of moment matching iterations ('max_iters' argument)
+    was reached.\n",
+    "Increasing the value may improve accuracy.",
+    call. = FALSE
+  )
+}
+
+#' Warning message if not using split transformation and accuracy is compromised
+#' @noRd
+throw_large_kf_warning <- function(kf) {
+  if (any(kf > 0.5)) {
```

Comment:
does it make sense to make the threshold an argument defaulting to 0.5?

Reply:
Changed